### PR TITLE
svg_loader: support other units for mask and clip 

### PR DESCRIPTION
**Pull request by mgrudzinska**
Added support for the objectBoundingBox value of
the clipPathUnits and for the userSpaceOnUse value
of the maskContentUnits.

@issue: https://github.com/thorvg/thorvg/issues/429

### CODE REVIEW

**Comment by mgrudzinska** [comment]
replacing #1547

File: src/loaders/svg/tvgSvgSceneBuilder.cpp [hunk 1]
```diff
@@ -92,6 +110,51 @@ static Box _boundingBox(const Shape* shape)
...
+    for (auto child : children) {
+        childBBox = _boundingBox(child);
+        _mergeBoxes(box, childBBox);
+    }
+
+    return box;
+}
+
+
+static Box _boundingBox(const Paint* paint)
```
Comment (hermet) [inline src/loaders/svg/tvgSvgSceneBuilder.cpp:141]: paint->bounds() will return the merged region of its internal elements. You are recomputing the bounds here. Is this necessary?

Also, It looks  "static Box _boundingBox(const Shape* shape)" can be removed?

File: src/loaders/svg/tvgSvgSceneBuilder.cpp [hunk 2]
```diff
@@ -292,9 +373,9 @@ static void _applyComposition(Paint* paint, const SvgNode* node, const Box& vBox
             node->style->mask.applying = true;
 
             bool isMaskWhite = true;
-            auto comp = _sceneBuildHelper(compNode, vBox, svgPath, true, 0, &isMaskWhite);
-            if (comp) {
-                if (node->transform) comp->transform(*node->transform);
+            if (auto comp = _sceneBuildHelper(compNode, vBox, svgPath, true, 0, &isMaskWhite)) {
+                Matrix finalTransform = _compositionTransform(paint, node, compNode, SvgNodeType::Mask);
+                if (!mathIdentity((const Matrix*)(&finalTransform))) comp->transform(finalTransform);
```
Comment (hermet) [inline src/loaders/svg/tvgSvgSceneBuilder.cpp:378]: --mathIdentity
https://github.com/thorvg/thorvg/pull/1603

File: src/loaders/svg/tvgSvgSceneBuilder.cpp [hunk 3]
```diff
@@ -269,13 +351,12 @@ static void _applyComposition(Paint* paint, const SvgNode* node, const Box& vBox
...
             }
 
-            if (node->transform) {
-                auto m = comp->transform();
-                m = mathMultiply(node->transform, &m);
-                comp->transform(m);
-            }
+            if (valid) {
+                Matrix finalTransform = _compositionTransform(paint, node, compNode, SvgNodeType::ClipPath);
+                if (!mathIdentity((const Matrix*)(&finalTransform))) comp->transform(finalTransform);
```
Comment (hermet) [inline src/loaders/svg/tvgSvgSceneBuilder.cpp:356]: --mathIdentity
https://github.com/thorvg/thorvg/pull/1603

**Review by hermet — COMMENTED** [review]
Thanks, Please check comments.

File: src/loaders/svg/tvgSvgSceneBuilder.cpp (hunk 1 above)
Comment (mgrudzinska) [reply src/loaders/svg/tvgSvgSceneBuilder.cpp:141]: I need to know the bounds without stroke. for a single shape I can remove it, but for a scene it's not possible, I need to traverse all elements. long time ago I wanted to change the bounds api accordingly, so I could use it here, but as I remember it wasn't approved.

Comment (hermet) [reply src/loaders/svg/tvgSvgSceneBuilder.cpp:141]: hmm but your logic looks including the stroking width as well.

Comment (hermet) [reply src/loaders/svg/tvgSvgSceneBuilder.cpp:141]: ah ok I mis-read the code.

**Comment by hermet** [comment]
This might be helpful:  https://github.com/thorvg/thorvg/pull/1608
```
#include "tvgPaint.h"
...
paint->pImpl->bounds(x, y,w, h, transform, stroking = false);
```

**Comment by mgrudzinska** [comment]
> This might be helpful: #1608
> 
> ```
> #include "tvgPaint.h"
> ...
> P(paint)->bounds(x, y,w, h, transform, stroking = false);
> ```


yes, I'll change this pr after #1608 is on the main

**Review by hermet — APPROVED** [review]
Thanks!

**Pull request merged**